src/main/scala/org/broadinstitute/dsde/vault/services/analysis/UpdateServiceHandler.scala

```diff
@@ -13,25 +14,57 @@ import spray.routing.RequestContext
 
 object UpdateServiceHandler {
 
-  case class UpdateMessage(dmId: String, update: AnalysisUpdate)
+  case class UpdateMessage(dmId: String, update: AnalysisUpdate, forceLocation: Option[String] = None)
 
-  def props(requestContext: RequestContext, dmService: ActorRef): Props =
-    Props(new UpdateServiceHandler(requestContext, dmService))
+  def props(requestContext: RequestContext, dmService: ActorRef, bossService: ActorRef): Props =
+    Props(new UpdateServiceHandler(requestContext, dmService, bossService))
 
 }
 
-case class UpdateServiceHandler(requestContext: RequestContext, dmService: ActorRef) extends Actor {
+case class UpdateServiceHandler(requestContext: RequestContext, dmService: ActorRef, bossService: ActorRef) extends Actor {
 
   implicit val system = context.system
   val log = Logging(system, getClass)
 
+  // the number of files associated with this analysis
+  // set by AnalysisUpdate
+  var fileCount: Integer = _
+
+  // a map from file types to Boss IDs
+  // added by BossObjectCreated messages (1 per message)
+  // consumed by DmClientService.DMUpdateAnalysis after [fileCount] have been added
+  var bossObjects: Map[String, String] = Map.empty
+
+  // a map from file types to Presigned Boss PUT URLs
+  // added by BossObjectResolved messages (1 per message)
+  // consumed by DMAnalysisUpdated after [fileCount] have been added and the DM ID has been set
+  var bossURLs: Map[String, String] = Map.empty
+
+  // the DM ID for this analysis
+  // set by UpdateMessage
+  var dmId: Option[String] = None
+
   override def receive: Receive = {
 
-    case UpdateMessage(dmId: String, update: AnalysisUpdate) =>
-      dmService ! DmClientService.DMUpdateAnalysis(dmId, update)
+    case UpdateMessage(id: String, update: AnalysisUpdate, forceLocation: Option[String]) =>
+      dmId = Some(id)
+      fileCount = update.files.size
+      update.files.foreach {
+        case (ftype, fpath) =>
+          bossService ! BossClientService.BossCreateObject(BossDefaults.getCreationRequest(fpath, forceLocation), ftype)
+      }
+
+    case BossObjectCreated(bossObject: BossCreationObject, creationKey: String) =>
+      bossObjects += creationKey -> bossObject.objectId.get
+      bossService ! BossClientService.BossResolveObject(BossDefaults.getResolutionRequest("PUT"), bossObject.objectId.get, creationKey)
+      if (fileCount == bossObjects.size)
+        dmService ! DmClientService.DMUpdateAnalysis(dmId.get, new AnalysisUpdate(bossObjects))
```

Comment:
take a look at the completeIfDone method in uBAM/IngestServiceHandler. I think you need something like that here. Otherwise, you have the possibility that you'll send DmClientService.DMUpdateAnalysis before all the boss urls have finished resolving.

Actually, looking at the code, I wonder if you could move the check on file count into BossObjectResolved. So, you'd have:
```
    case BossObjectCreated(bossObject: BossCreationObject, creationKey: String) =>
      bossObjects += creationKey -> bossObject.objectId.get
      bossService ! BossClientService.BossResolveObject(BossDefaults.getResolutionRequest("PUT"), bossObject.objectId.get, creationKey)

    case BossObjectResolved(bossObject: BossResolutionResponse, creationKey: String) =>
      bossURLs += creationKey -> bossObject.objectUrl
      if (fileCount == bossURLs.size)
        dmService ! DmClientService.DMUpdateAnalysis(dmId.get, new AnalysisUpdate(bossObjects))
```